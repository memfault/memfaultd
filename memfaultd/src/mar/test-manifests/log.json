{
    "schema_version": 1,
    "collection_time": {
        "timestamp": "1970-01-01T00:00:04.321Z",
        "uptime_ms": 800000,
        "elapsed_realtime_ms": 900000,
        "linux_boot_id": "230295cb-04d4-40b8-8624-ec37089b9b75",
        "boot_count": 0
    },
    "type": "linux-logs",
    "device": {
        "project_key": "INSERT_PROJECT_KEY_HERE",
        "hardware_version": "evt",
        "software_version": "1.0.0",
        "software_type": "linux-build",
        "device_serial": "DEMOSERIAL"
    },
    "metadata": {
        "format": { "id": "v1", "serialization": "json-lines" },
        "producer": { "id": "memfaultd", "version": "2.0.0" },
        "log_file_name": "logs.txt",
        "cid": { "uuid": "271e5249-1e3f-4c89-b4a5-ed96cc8f42a7" },
        "next_cid": { "uuid": "4276a26b-ba89-4c57-832f-d9e90e9629dd" }
    }
}
