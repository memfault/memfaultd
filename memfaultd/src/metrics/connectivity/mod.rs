//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod connectivity_monitor;
pub use connectivity_monitor::ConnectivityMonitor;

mod report_sync_event_handler;
pub use report_sync_event_handler::ReportSyncEventHandler;
