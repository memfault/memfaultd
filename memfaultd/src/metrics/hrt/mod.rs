//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod report;
pub use report::{HrtReport, HRT_DEFAULT_MAX_SAMPLES_PER_MIN};

mod schema;
pub use schema::write_report_to_disk;
// Export for use in tests
#[cfg(test)]
pub use schema::HighResTelemetryV1;
