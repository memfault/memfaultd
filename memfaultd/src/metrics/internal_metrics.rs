//
// Copyright (c) Memfault, Inc.
// See License.txt for details
// MAR cleaner metrics
pub const INTERNAL_METRIC_MAR_ENTRY_COUNT: &str = "MemfaultSdkMetric_mar_entry_count";
pub const INTERNAL_METRIC_MAR_CLEANER_DURATION: &str =
    "MemfaultSdkMetric_mar_clean_duration_seconds";
pub const INTERNAL_METRIC_MAR_ENTRIES_DELETED: &str = "MemfaultSdkMetric_mar_entries_deleted";
