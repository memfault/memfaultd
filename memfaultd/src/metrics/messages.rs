//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::path::PathBuf;

use eyre::Result;
use ssf::{Message, MsgMailbox};

use crate::network::NetworkConfig;

use super::{KeyedMetricReading, MetricReportType, SessionName};

/// Allows KeyedMetricReading to be sent as a message. The `ssf` framework will
/// automatically support sending `Vec<KeyedMetricReading>` as well.
impl Message for KeyedMetricReading {
    type Reply = Result<()>;
}

/// Syntactic Sugar because this is the most-used type of mailbox in memfaultd.
pub type MetricsMBox = MsgMailbox<Vec<KeyedMetricReading>>;

pub enum SessionEventMessage {
    StartSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
    },
    StopSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
        mar_staging_area: PathBuf,
        network_config: NetworkConfig,
    },
}

impl Message for SessionEventMessage {
    type Reply = Result<()>;
}

#[derive(Clone)]
pub enum ReportsToDump {
    All,
    Report(MetricReportType),
}

#[derive(Clone)]
pub struct DumpMetricReportMessage {
    reports_to_dump: ReportsToDump,
    mar_staging_area: PathBuf,
    network_config: NetworkConfig,
}

impl DumpMetricReportMessage {
    pub fn new(
        reports_to_dump: ReportsToDump,
        mar_staging_area: PathBuf,
        network_config: NetworkConfig,
    ) -> Self {
        Self {
            reports_to_dump,
            mar_staging_area,
            network_config,
        }
    }

    pub fn mar_staging_area(&self) -> &PathBuf {
        &self.mar_staging_area
    }

    pub fn network_config(&self) -> &NetworkConfig {
        &self.network_config
    }

    pub fn reports_to_dump(&self) -> &ReportsToDump {
        &self.reports_to_dump
    }
}
impl Message for DumpMetricReportMessage {
    type Reply = Result<()>;
}

#[derive(Clone)]
pub struct DumpHrtMessage {
    mar_staging_area: PathBuf,
    network_config: NetworkConfig,
}

impl DumpHrtMessage {
    pub fn new(mar_staging_area: PathBuf, network_config: NetworkConfig) -> Self {
        Self {
            mar_staging_area,
            network_config,
        }
    }

    pub fn mar_staging_area(&self) -> &PathBuf {
        &self.mar_staging_area
    }

    pub fn network_config(&self) -> &NetworkConfig {
        &self.network_config
    }
}

impl Message for DumpHrtMessage {
    type Reply = Result<()>;
}
