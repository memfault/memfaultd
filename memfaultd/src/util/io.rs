//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::{
    self, copy, sink, BufReader, Chain, Cursor, ErrorKind, IoSlice, Read, Seek, SeekFrom, Write,
};

/// A trait for getting the length of a stream.
/// Note std::io::Seek also has a stream_len() method, but that method is fallible.
pub trait StreamLen {
    /// Gets the length of the stream in bytes.
    fn stream_len(&self) -> u64;
}

impl<R: StreamLen> StreamLen for BufReader<R> {
    fn stream_len(&self) -> u64 {
        self.get_ref().stream_len()
    }
}

impl<A: AsRef<[u8]>, B: StreamLen> StreamLen for Chain<Cursor<A>, B> {
    fn stream_len(&self) -> u64 {
        let (a, b) = self.get_ref();

        a.get_ref().as_ref().len() as u64 + b.stream_len()
    }
}

pub trait StreamPosition {
    /// Gets the current position in the stream.
    fn stream_position(&self) -> usize;
}

/// A passthrough `Write` implementation that keeps track of position.
///
/// This is used to keep track of the current position in the output stream, since we can't use
/// `Seek` on all output streams. Additionally this allows us to keep track of the position
/// when using functions like `copy` that may call write several times and potentially fail.
pub struct StreamPositionTracker<T: Write> {
    writer: T,
    pos: usize,
}

impl<T: Write> StreamPositionTracker<T> {
    pub fn new(writer: T) -> Self {
        Self { writer, pos: 0 }
    }
}

impl<T: Write> Write for StreamPositionTracker<T> {
    /// Passthrough to the underlying writer, but also updates the position.
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        let written = self.writer.write(buf)?;
        self.pos += written;
        Ok(written)
    }

    // Passthrough to the underlying writer.
    fn flush(&mut self) -> io::Result<()> {
        self.writer.flush()
    }

    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> std::io::Result<usize> {
        let written = self.writer.write_vectored(bufs)?;
        self.pos += written;
        Ok(written)
    }
}

impl<T: Write> StreamPosition for StreamPositionTracker<T> {
    fn stream_position(&self) -> usize {
        self.pos
    }
}

/// Stream wrapper that implements a subset of `Seek`.
///
/// This is needed because we read from streams that do not implement `Seek`,
/// such as `Stdin`. Instead of implementing ad-hoc skipping we'll implement
/// `Seek` such that it only allows seeking forward. Any seek from the end of
/// the stream, or that would go backwards, will result in an error.
pub struct ForwardOnlySeeker<T: Read> {
    reader: T,
    pos: u64,
}

impl<T: Read> ForwardOnlySeeker<T> {
    pub fn new(reader: T) -> Self {
        Self { reader, pos: 0 }
    }
}

impl<T: Read> Read for ForwardOnlySeeker<T> {
    /// Passthrough to the underlying reader, but also updates the position.
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let bytes_read = self.reader.read(buf)?;
        self.pos += bytes_read as u64;
        Ok(bytes_read)
    }
}

impl<T: Read> Seek for ForwardOnlySeeker<T> {
    /// Seeks forward in the stream, returns an error if seeking backwards.
    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
        let seek_size = match pos {
            SeekFrom::Current(offset) => u64::try_from(offset).ok(),
            SeekFrom::Start(offset) => offset.checked_sub(self.pos),
            SeekFrom::End(_) => {
                return Err(io::Error::new(
                    ErrorKind::InvalidInput,
                    "Cannot seek from end",
                ))
            }
        };

        match seek_size {
            Some(seek_size) => {
                copy(&mut self.by_ref().take(seek_size), &mut sink())?;
                Ok(self.pos)
            }
            None => Err(io::Error::new(
                ErrorKind::InvalidInput,
                "Only seeking forward allowed",
            )),
        }
    }

    fn stream_position(&mut self) -> io::Result<u64> {
        Ok(self.pos)
    }
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    use super::*;

    #[test]
    fn test_write_cursor() {
        let mut buf = Vec::new();
        let mut writer = StreamPositionTracker::new(&mut buf);

        writer.write_all(b"hello").unwrap();
        assert_eq!(writer.stream_position(), 5);
        writer.write_all(b" world").unwrap();
        assert_eq!(writer.stream_position(), 11);
        writer.write_all(b"!").unwrap();
        assert_eq!(writer.stream_position(), 12);

        assert_eq!(buf, b"hello world!");
    }

    #[test]
    fn test_write_vectored_cursor() {
        let mut buf = Vec::new();
        let mut writer = StreamPositionTracker::new(&mut buf);

        let write_vector = [
            IoSlice::new(b"hello"),
            IoSlice::new(b" world"),
            IoSlice::new(b"!"),
        ];
        let bytes_written = writer.write_vectored(&write_vector).unwrap();

        assert_eq!(bytes_written, 12);
        assert_eq!(writer.stream_position(), 12);
        assert_eq!(buf, b"hello world!");
    }

    #[test]
    fn test_forward_seeker_stream() {
        let mut input_stream = b"Hello world".as_ref();

        let mut reader = ForwardOnlySeeker::new(&mut input_stream);
        let mut out_buf = [0u8; 5];
        reader.read_exact(&mut out_buf).unwrap();
        assert_eq!(&out_buf, b"Hello");

        reader.seek(SeekFrom::Current(1)).unwrap();
        reader.read_exact(&mut out_buf).unwrap();
        assert_eq!(&out_buf, b"world");
    }

    #[rstest]
    #[case(SeekFrom::End(0))]
    #[case(SeekFrom::Current(-1))]
    #[case(SeekFrom::Start(0))]
    fn test_forward_seeker_seek_fail(#[case] seek: SeekFrom) {
        let mut reader = ForwardOnlySeeker::new(b"Hello world".as_ref());
        reader.seek(SeekFrom::Start(1)).unwrap();
        assert!(reader.seek(seek).is_err());
    }
}
