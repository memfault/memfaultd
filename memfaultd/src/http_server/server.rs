//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{net::SocketAddr, thread::spawn};

use eyre::{eyre, Result};
use log::{debug, trace, warn};
use tiny_http::{Request, Response, Server};

use crate::http_server::{HttpHandler, HttpHandlerResult};

/// A server that listens for collectd JSON pushes and stores them in memory.
pub struct HttpServer {
    handlers: Option<Vec<Box<dyn HttpHandler>>>,
}

impl HttpServer {
    pub fn new(handlers: Vec<Box<dyn HttpHandler>>) -> Self {
        HttpServer {
            handlers: Some(handlers),
        }
    }

    pub fn start(&mut self, listening_address: SocketAddr) -> Result<()> {
        let server = Server::http(listening_address).map_err(|e| {
            eyre!("Error starting server: could not bind to {listening_address}: {e}")
        })?;

        if let Some(handlers) = self.handlers.take() {
            spawn(move || {
                debug!("HTTP Server started on {listening_address}");

                for request in server.incoming_requests() {
                    Self::handle_request(&handlers, request);
                }
            });
            Ok(())
        } else {
            Err(eyre!("HTTP Server already started"))
        }
    }

    fn handle_request(handlers: &[Box<dyn HttpHandler>], mut request: Request) {
        trace!(
            "HTTP request {:?} {:?}\n{:?}",
            request.method(),
            request.url(),
            request.headers()
        );

        let method = request.method().to_owned();
        let url = request.url().to_owned();
        for handler in handlers.iter() {
            match handler.handle_request(&mut request) {
                HttpHandlerResult::Response(response) => {
                    if let Err(e) = request.respond(response) {
                        warn!("HTTP: Error sending response {} {}: {:?}", method, url, e);
                    }
                    return;
                }
                HttpHandlerResult::Error(e) => {
                    warn!("HTTP: Error processing request {} {}: {}", method, url, e);
                    let _r = request
                        .respond(Response::empty(500).with_data(e.as_bytes(), Some(e.len())));
                    return;
                }
                HttpHandlerResult::NotHandled => { /* continue  */ }
            };
        }
        debug!("HTTP[404] {} {}", method, url);
        let _r = request.respond(Response::empty(404));
    }
}
