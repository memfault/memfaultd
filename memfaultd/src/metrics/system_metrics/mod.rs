//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    collections::HashSet,
    path::PathBuf,
    thread::sleep,
    time::{Duration, Instant},
};

use disk::{get_tracked_disks, DiskMetricsCollector, DISK_METRIC_NAMESPACE};
use eyre::{eyre, Result};
use log::{debug, error};

use crate::{
    metrics::KeyedMetricReading,
    mmc::MmcImpl,
    util::system::{bytes_per_page, clock_ticks_per_second},
};

mod cpu;
use crate::metrics::system_metrics::cpu::{CpuMetricCollector, CPU_METRIC_NAMESPACE};

mod thermal;
use thermal::ThermalMetricsCollector;
pub use thermal::THERMAL_METRIC_NAMESPACE;

mod memory;
use crate::metrics::system_metrics::memory::{MemoryMetricsCollector, MEMORY_METRIC_NAMESPACE};

mod network_interfaces;
use network_interfaces::{NetworkInterfaceMetricCollector, NetworkInterfaceMetricsConfig};
pub use network_interfaces::{
    METRIC_INTERFACE_BYTES_PER_SECOND_RX_SUFFIX, METRIC_INTERFACE_BYTES_PER_SECOND_TX_SUFFIX,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
};

mod processes;
pub use processes::ProcessMetricsConfig;
use processes::{ProcessMetricsCollector, PROCESSES_METRIC_NAMESPACE};

mod disk;

mod disk_space;
pub use disk_space::DiskSpaceMetricsConfig;
use disk_space::{
    DiskSpaceMetricCollector, NixStatvfs, DISKSPACE_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
};

mod diskstats;
pub use diskstats::DiskstatsMetricsConfig;
use diskstats::{DiskstatsMetricCollector, DISKSTATS_METRIC_NAMESPACE};

use self::{
    memory::{MemInfoParser, MemInfoParserImpl},
    processes::ProcfsProcessNameMapper,
};
use super::MetricsMBox;

pub const BUILTIN_SYSTEM_METRIC_NAMESPACES: &[&str; 9] = &[
    CPU_METRIC_NAMESPACE,
    MEMORY_METRIC_NAMESPACE,
    THERMAL_METRIC_NAMESPACE,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
    PROCESSES_METRIC_NAMESPACE,
    DISK_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE,
    DISKSTATS_METRIC_NAMESPACE,
    // Include in list of namespaces so that
    // legacy collectd from the "df" plugin
    // are still filtered out
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
];

pub trait SystemMetricFamilyCollector {
    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>>;
    fn family_name(&self) -> &'static str;
}
pub struct SystemMetricsCollector {
    metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>>,
    metrics_mbox: MetricsMBox,
}

impl SystemMetricsCollector {
    pub fn new(
        processes_config: ProcessMetricsConfig,
        network_interfaces_config: Option<HashSet<String>>,
        disk_space_config: DiskSpaceMetricsConfig,
        diskstats_config: DiskstatsMetricsConfig,
        metrics_mbox: MetricsMBox,
    ) -> Self {
        // CPU, Memory, and Thermal metrics are captured by default
        let mut metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>> = vec![
            Box::new(CpuMetricCollector::new()),
            Box::new(MemoryMetricsCollector::new(MemInfoParserImpl::new())),
            Box::new(ThermalMetricsCollector::new()),
        ];

        // Check if process metrics have been manually configured
        match processes_config {
            // Monitoring no processes means this collector is disabled
            ProcessMetricsConfig::Processes(processes) if processes.is_empty() => {}
            // In all other cases we can just directly pass the config to ProcessMetricsCollector
            process_metrics_config => {
                // We need the total memory for the system to calculate the
                // percent used by each individual process
                if let Ok(mem_total) = Self::get_total_memory() {
                    metric_family_collectors.push(Box::new(ProcessMetricsCollector::<
                        Instant,
                        ProcfsProcessNameMapper,
                    >::new(
                        process_metrics_config,
                        clock_ticks_per_second() as f64 / 1000.0,
                        bytes_per_page() as f64,
                        mem_total,
                    )))
                }
            }
        };

        // Check if disk space metrics have been manually configured
        match disk_space_config {
            // Monitoring no disks means this collector is disabled
            DiskSpaceMetricsConfig::Disks(disks) if disks.is_empty() => {}
            disk_space_metrics_config => metric_family_collectors.push(Box::new(
                DiskSpaceMetricCollector::new(NixStatvfs::new(), disk_space_metrics_config),
            )),
        };

        // Check if diskstats metrics have been manually configured
        match diskstats_config.clone() {
            // Monitoring no devices means this collector is disabled
            DiskstatsMetricsConfig::Devices(devices) if devices.is_empty() => {}
            diskstats_config => metric_family_collectors.push(Box::new(
                DiskstatsMetricCollector::<Instant>::new(diskstats_config),
            )),
        };

        // TODO: Implement actual config values for disk metrics
        match diskstats_config {
            // Monitoring no devices means this collector is disabled
            DiskstatsMetricsConfig::Devices(devices) if devices.is_empty() => {}
            diskstats_config => match get_tracked_disks(diskstats_config, "/sys/block") {
                Ok(disks) => {
                    let mmc = disks
                        .into_iter()
                        .filter_map(|disk_path| match MmcImpl::new(PathBuf::from(disk_path)) {
                            Ok(mmc) => Some(mmc),
                            Err(e) => {
                                debug!("Failed to open disk: {}", e);
                                None
                            }
                        })
                        .collect::<Vec<_>>();
                    metric_family_collectors.push(Box::new(DiskMetricsCollector::new(mmc)));
                }
                Err(e) => error!("Failed to start disk metrics collector: {}", e),
            },
        };

        // Check if network interface metrics have been manually configured
        match network_interfaces_config {
            Some(interfaces) if !interfaces.is_empty() => metric_family_collectors.push(Box::new(
                NetworkInterfaceMetricCollector::<Instant>::new(
                    NetworkInterfaceMetricsConfig::Interfaces(interfaces),
                ),
            )),
            // Monitoring no interfaces means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(NetworkInterfaceMetricCollector::<
                Instant,
            >::new(
                NetworkInterfaceMetricsConfig::Auto
            ))),
        };

        Self {
            metric_family_collectors,
            metrics_mbox,
        }
    }

    fn get_total_memory() -> Result<f64> {
        let mem_info_parser = MemInfoParserImpl::new();
        let mut stats = mem_info_parser.get_meminfo_stats()?;
        stats
            .remove("MemTotal")
            .ok_or_else(|| eyre!("Couldn't get MemTotal"))
    }

    pub fn run(&mut self, metric_poll_duration: Duration) {
        loop {
            for collector in self.metric_family_collectors.iter_mut() {
                match collector.collect_metrics() {
                    Ok(readings) => {
                        if let Err(e) = self.metrics_mbox.send_and_forget(readings) {
                            debug!(
                                "Couldn't add metric reading for family \"{}\": {:?}",
                                collector.family_name(),
                                e
                            )
                        }
                    }
                    Err(e) => debug!(
                        "Failed to collect readings for family \"{}\": {}",
                        collector.family_name(),
                        e
                    ),
                }
            }

            sleep(metric_poll_duration);
        }
    }
}
