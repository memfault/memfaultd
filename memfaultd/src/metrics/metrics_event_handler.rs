//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::Read;

use crate::http_server::{HttpHandler, HttpHandlerResult, MetricsRequest};
use eyre::Result;
use log::error;
use tiny_http::{Method, Request, Response};

use super::MetricsMBox;

pub struct MetricsEventHandler {
    metrics_mbox: MetricsMBox,
    data_collection_enabled: bool,
}

impl MetricsEventHandler {
    pub fn new(metrics_mbox: MetricsMBox, data_collection_enabled: bool) -> Self {
        Self {
            metrics_mbox,
            data_collection_enabled,
        }
    }

    pub fn parse_request<R: Read>(&self, reader: &mut R) -> Result<MetricsRequest> {
        let mut buf = Vec::new();
        reader.read_to_end(&mut buf)?;
        let request: MetricsRequest = serde_json::from_slice(&buf)?;
        Ok(request)
    }
}

impl HttpHandler for MetricsEventHandler {
    fn handle_request(&self, request: &mut Request) -> HttpHandlerResult {
        if request.url() != "/v1/metrics" || *request.method() != Method::Post {
            return HttpHandlerResult::NotHandled;
        }

        if self.data_collection_enabled {
            match self.parse_request(&mut request.as_reader()) {
                Ok(metrics_request) => {
                    match self.metrics_mbox.send_and_forget(metrics_request.readings) {
                        Ok(()) => HttpHandlerResult::Response(Response::empty(200).boxed()),
                        Err(e) => {
                            error!("Failed to send metrics to mailbox: {}", e);
                            HttpHandlerResult::Response(Response::empty(500).boxed())
                        }
                    }
                }
                Err(e) => {
                    error!("Failed to parse metrics request: {}", e);
                    HttpHandlerResult::Response(Response::empty(400).boxed())
                }
            }
        } else {
            HttpHandlerResult::Response(Response::empty(200).boxed())
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::metrics::{KeyedMetricReading, MetricReportManager, MetricStringKey, MetricValue};
    use crate::test_utils::setup_logger;

    use std::{collections::HashMap, io::Cursor};

    use insta::{assert_json_snapshot, with_settings};
    use rstest::{fixture, rstest};
    use ssf::ServiceJig;
    use tiny_http::TestRequest;

    const TEST_METRICS_BODY: &str = "
    {
        \"readings\": [
                {\"name\": \"Daisy\", \"value\": {\"Gauge\": {\"value\": 1.0, \"timestamp\": \"2024-01-01 00:00:00 UTC\"}}},
                {\"name\": \"Carlton\", \"value\": {\"Gauge\": {\"value\": 2.0, \"timestamp\": \"2024-01-01 00:00:00 UTC\"}}}
        ]
    }";

    #[rstest]
    fn test_parse_request(mut fixture: Fixture) {
        let handler = &mut fixture.handler;

        let key = MetricStringKey::from("Daisy");
        let reading = KeyedMetricReading::new_histogram(key.clone(), 42.0);
        let request = MetricsRequest {
            readings: vec![reading],
        };

        let request_json = serde_json::to_string(&request).unwrap();
        let mut cursor = Cursor::new(request_json.as_bytes());
        let parsed_request = handler.parse_request(&mut cursor).unwrap();

        assert_eq!(parsed_request.readings.len(), 1);
        assert_eq!(parsed_request.readings[0].name, key);
    }

    #[rstest]
    fn test_handle_request(mut fixture: Fixture, _setup_logger: ()) {
        let handler = &mut fixture.handler;

        let request = TestRequest::new()
            .with_path("/v1/metrics")
            .with_method(Method::Post)
            .with_body(TEST_METRICS_BODY);

        let response = handler.handle_request(&mut request.into());
        match response {
            HttpHandlerResult::Response(response) => {
                assert_eq!(response.status_code(), 200);
            }
            _ => panic!("Unexpected response"),
        }
        fixture.process_all();

        let metrics = fixture.take_metrics();
        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(metrics);
        });
    }

    #[rstest]
    fn test_handle_request_data_collection_disabled(mut fixture: Fixture, _setup_logger: ()) {
        fixture.set_data_collection_enabled(false);
        let handler = &mut fixture.handler;

        let request = TestRequest::new()
            .with_path("/v1/metrics")
            .with_method(Method::Post)
            .with_body(TEST_METRICS_BODY);

        let response = handler.handle_request(&mut request.into());
        match response {
            HttpHandlerResult::Response(response) => {
                assert_eq!(response.status_code(), 200);
            }
            _ => panic!("Unexpected response"),
        }
        fixture.process_all();

        let metrics = fixture.take_metrics();
        assert!(metrics.is_empty());
    }

    struct Fixture {
        handler: MetricsEventHandler,
        service: ServiceJig<MetricReportManager>,
    }

    impl Fixture {
        fn take_metrics(&mut self) -> HashMap<MetricStringKey, MetricValue> {
            self.service.get_service_mut().take_heartbeat_metrics()
        }

        fn process_all(&mut self) {
            self.service.process_all();
        }

        fn set_data_collection_enabled(&mut self, enabled: bool) {
            self.handler.data_collection_enabled = enabled;
        }
    }

    #[fixture]
    fn fixture() -> Fixture {
        let report_manager = MetricReportManager::default();
        let service = ServiceJig::prepare(report_manager);
        let handler = MetricsEventHandler::new(service.mailbox.clone().into(), true);

        Fixture { handler, service }
    }
}
