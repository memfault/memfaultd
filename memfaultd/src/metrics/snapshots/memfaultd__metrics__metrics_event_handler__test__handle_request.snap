---
source: memfaultd/src/metrics/metrics_event_handler.rs
expression: metrics
snapshot_kind: text
---
{
  "Carlton": 2.0,
  "Daisy": 1.0
}
